File: csharp/Oojs/Validator.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Oojs;

public sealed class Validator
{
    private readonly Registry _registry;
    private readonly bool _failFast;

    public Validator(Registry registry, bool failFast = false)
    {
        _registry = registry;
        _failFast = failFast;
    }

    public List<ValidationError> Validate(object? instance, TypeDef targetType, Schema schema)
    {
        var errors = new List<ValidationError>();
        var normalized = JsonUtil.Normalize(instance);
        ValidateInstance(normalized, targetType, schema, "/", errors);
        return errors;
    }

    public List<ValidationError> ValidateJson(string text, string targetTypeName, Schema schema)
    {
        using var doc = JsonDocument.Parse(text);
        var instance = JsonUtil.FromJson(doc.RootElement);
        if (!schema.Types.TryGetValue(targetTypeName, out var typedef))
        {
            throw new ArgumentException(
                $"Type '{targetTypeName}' not found in schema '{schema.SchemaId}'"
            );
        }
        return Validate(instance, typedef, schema);
    }

    private bool ValidateInstance(object? instance, TypeDef target, Schema schema, string path, List<ValidationError> errors)
    {
        if (instance is not Dictionary<string, object?> obj)
        {
            errors.Add(new ValidationError(path, ErrorCode.TYPE_MISMATCH, "expected a JSON object"));
            return !_failFast;
        }

        var discName = schema.Discriminator;
        if (!obj.TryGetValue(discName, out var discValObj))
        {
            errors.Add(new ValidationError(path, ErrorCode.MISSING_DISCRIMINATOR,
                $"missing discriminator property '{discName}'"));
            return !_failFast;
        }

        if (discValObj is not string discVal)
        {
            errors.Add(new ValidationError($"{path}/{discName}", ErrorCode.INVALID_DISCRIMINATOR_TYPE,
                $"discriminator property '{discName}' must be a string"));
            return !_failFast;
        }

        var concrete = _registry.LookupByDiscriminatorValue(discVal);
        if (concrete is null)
        {
            errors.Add(new ValidationError($"{path}/{discName}", ErrorCode.UNKNOWN_TYPE,
                $"unknown type '{discVal}'"));
            return !_failFast;
        }

        if (concrete.Abstract)
        {
            errors.Add(new ValidationError($"{path}/{discName}", ErrorCode.ABSTRACT_TYPE,
                $"type '{discVal}' is abstract and cannot be instantiated"));
            return !_failFast;
        }

        if (!concrete.IsSubtypeOf(target))
        {
            errors.Add(new ValidationError($"{path}/{discName}", ErrorCode.TYPE_MISMATCH,
                $"type '{discVal}' is not a subtype of '{target.Name}'"));
            return !_failFast;
        }

        foreach (var req in concrete.EffectiveRequired())
        {
            if (!obj.ContainsKey(req))
            {
                errors.Add(new ValidationError(path, ErrorCode.MISSING_REQUIRED,
                    $"missing required property '{req}'"));
                if (_failFast)
                {
                    return false;
                }
            }
        }

        var effectiveProps = concrete.EffectiveProperties();
        foreach (var kvp in obj)
        {
            if (kvp.Key == discName)
            {
                continue;
            }
            if (!effectiveProps.TryGetValue(kvp.Key, out var propDef))
            {
                if (schema.ClosedWorld)
                {
                    errors.Add(new ValidationError($"{path}/{Escape(kvp.Key)}", ErrorCode.ADDITIONAL_PROPERTY,
                        $"unexpected additional property '{kvp.Key}'"));
                    if (_failFast)
                    {
                        return false;
                    }
                }
                continue;
            }
            var ok = ValidateProperty(kvp.Value, propDef, schema, $"{path}/{Escape(kvp.Key)}", errors);
            if (!ok && _failFast)
            {
                return false;
            }
        }

        return true;
    }

    private bool ValidateProperty(object? value, PropertyDef prop, Schema schema, string path, List<ValidationError> errors)
    {
        if (prop is ArrayProperty arr)
        {
            return ValidateArray(value, arr, schema, path, errors);
        }
        if (prop is PrimitiveProperty prim)
        {
            return ValidatePrimitive(value, prim, path, errors);
        }
        if (prop is TypeRefProperty typeRef)
        {
            return ValidateTypeRef(value, typeRef, schema, path, errors);
        }
        return true;
    }

    private bool ValidateArray(object? value, ArrayProperty prop, Schema schema, string path, List<ValidationError> errors)
    {
        if (value is not List<object?> list)
        {
            errors.Add(new ValidationError(path, ErrorCode.TYPE_MISMATCH, "expected an array"));
            return !_failFast;
        }

        if (list.Count < prop.MinItems)
        {
            errors.Add(new ValidationError(path, ErrorCode.ARRAY_TOO_SHORT,
                $"array has {list.Count} item(s), minimum is {prop.MinItems}"));
            if (_failFast)
            {
                return false;
            }
        }

        if (prop.MaxItems is not null && list.Count > prop.MaxItems)
        {
            errors.Add(new ValidationError(path, ErrorCode.ARRAY_TOO_LONG,
                $"array has {list.Count} item(s), maximum is {prop.MaxItems}"));
            if (_failFast)
            {
                return false;
            }
        }

        if (prop.UniqueItems)
        {
            var seen = new HashSet<string>(StringComparer.Ordinal);
            foreach (var item in list)
            {
                var key = SerialiseForUniqueness(item);
                if (!seen.Add(key))
                {
                    errors.Add(new ValidationError(path, ErrorCode.ARRAY_DUPLICATE_ITEMS,
                        "array items must be unique"));
                    if (_failFast)
                    {
                        return false;
                    }
                    break;
                }
            }
        }

        for (var i = 0; i < list.Count; i++)
        {
            var ok = ValidateProperty(list[i], prop.Items, schema, $"{path}/{i}", errors);
            if (!ok && _failFast)
            {
                return false;
            }
        }

        return true;
    }

    private bool ValidatePrimitive(object? value, PrimitiveProperty prop, string path, List<ValidationError> errors)
    {
        if (!JsonKindMatches(value, prop.Kind))
        {
            errors.Add(new ValidationError(path, ErrorCode.TYPE_MISMATCH,
                $"expected {prop.Kind}, got {JsonTypeName(value)}"));
            return !_failFast;
        }

        if (prop.Kind == "integer" && value is double d && Math.Abs(d % 1) > 1e-12)
        {
            errors.Add(new ValidationError(path, ErrorCode.NOT_INTEGER,
                "value must be an integer (no fractional part)"));
            if (_failFast)
            {
                return false;
            }
        }

        if (prop.Kind == "string" && value is string s)
        {
            var length = s.Length;
            if (prop.MinLength is not null && length < prop.MinLength)
            {
                errors.Add(new ValidationError(path, ErrorCode.STRING_TOO_SHORT,
                    $"string length {length} < minLength {prop.MinLength}"));
                if (_failFast)
                {
                    return false;
                }
            }
            if (prop.MaxLength is not null && length > prop.MaxLength)
            {
                errors.Add(new ValidationError(path, ErrorCode.STRING_TOO_LONG,
                    $"string length {length} > maxLength {prop.MaxLength}"));
                if (_failFast)
                {
                    return false;
                }
            }
            if (prop.Pattern is not null)
            {
                try
                {
                    var re = new Regex(prop.Pattern);
                    if (!re.IsMatch(s))
                    {
                        errors.Add(new ValidationError(path, ErrorCode.PATTERN_MISMATCH,
                            $"value does not match pattern '{prop.Pattern}'"));
                        if (_failFast)
                        {
                            return false;
                        }
                    }
                }
                catch (ArgumentException e)
                {
                    errors.Add(new ValidationError(path, ErrorCode.PATTERN_MISMATCH,
                        $"invalid pattern '{prop.Pattern}': {e.Message}"));
                    if (_failFast)
                    {
                        return false;
                    }
                }
            }
            if (prop.Enum is not null && !EnumContains(prop.Enum, s))
            {
                errors.Add(new ValidationError(path, ErrorCode.ENUM_MISMATCH,
                    $"value '{s}' not in enum [{string.Join(",", prop.Enum)}]"));
                if (_failFast)
                {
                    return false;
                }
            }
        }
        else if ((prop.Kind == "integer" || prop.Kind == "number") && value is not null)
        {
            if (value is not long && value is not int && value is not double && value is not float)
            {
                return !_failFast;
            }
            var num = JsonUtil.AsDouble(value);
            if (prop.Minimum is not null && num < prop.Minimum)
            {
                errors.Add(new ValidationError(path, ErrorCode.BELOW_MINIMUM,
                    $"value {value} < minimum {prop.Minimum}"));
                if (_failFast)
                {
                    return false;
                }
            }
            if (prop.Maximum is not null && num > prop.Maximum)
            {
                errors.Add(new ValidationError(path, ErrorCode.ABOVE_MAXIMUM,
                    $"value {value} > maximum {prop.Maximum}"));
                if (_failFast)
                {
                    return false;
                }
            }
            if (prop.ExclusiveMinimum is not null && num <= prop.ExclusiveMinimum)
            {
                errors.Add(new ValidationError(path, ErrorCode.BELOW_EXCLUSIVE_MINIMUM,
                    $"value {value} must be > {prop.ExclusiveMinimum}"));
                if (_failFast)
                {
                    return false;
                }
            }
            if (prop.ExclusiveMaximum is not null && num >= prop.ExclusiveMaximum)
            {
                errors.Add(new ValidationError(path, ErrorCode.ABOVE_EXCLUSIVE_MAXIMUM,
                    $"value {value} must be < {prop.ExclusiveMaximum}"));
                if (_failFast)
                {
                    return false;
                }
            }
            if (prop.MultipleOf is not null)
            {
                var remainder = Math.Abs(num % prop.MultipleOf.Value);
                var rounded = Math.Round(remainder, 10);
                if (rounded != 0.0 && rounded != prop.MultipleOf.Value)
                {
                    errors.Add(new ValidationError(path, ErrorCode.NOT_MULTIPLE_OF,
                        $"value {value} is not a multiple of {prop.MultipleOf}"));
                    if (_failFast)
                    {
                        return false;
                    }
                }
            }
            if (prop.Enum is not null && !EnumContains(prop.Enum, value))
            {
                errors.Add(new ValidationError(path, ErrorCode.ENUM_MISMATCH,
                    $"value {value} not in enum [{string.Join(",", prop.Enum)}]"));
                if (_failFast)
                {
                    return false;
                }
            }
        }

        return true;
    }

    private bool ValidateTypeRef(object? value, TypeRefProperty prop, Schema schema, string path, List<ValidationError> errors)
    {
        if (value is not Dictionary<string, object?>)
        {
            errors.Add(new ValidationError(path, ErrorCode.TYPE_MISMATCH,
                "expected a JSON object for type reference"));
            return !_failFast;
        }

        var refType = prop.ResolvedType;
        if (refType is null)
        {
            errors.Add(new ValidationError(path, ErrorCode.UNKNOWN_TYPE,
                $"type reference '{prop.TypeName}' was not resolved at load time"));
            return !_failFast;
        }

        var refSchema = _registry.GetSchema(refType.SchemaId) ?? schema;
        return ValidateInstance(value, refType, refSchema, path, errors);
    }

    private static bool JsonKindMatches(object? value, string kind)
    {
        return kind switch
        {
            "null" => value is null,
            "boolean" => value is bool,
            "string" => value is string,
            "integer" => value is long || value is int || value is double || value is float,
            "number" => value is long || value is int || value is double || value is float,
            _ => false,
        };
    }

    private static string JsonTypeName(object? value)
    {
        return value switch
        {
            null => "null",
            bool => "boolean",
            string => "string",
            long => "integer",
            int => "integer",
            double => "number",
            float => "number",
            List<object?> => "array",
            Dictionary<string, object?> => "object",
            _ => value.GetType().Name,
        };
    }

    private static string Escape(string token) => token.Replace("~", "~0").Replace("/", "~1");

    private static string SerialiseForUniqueness(object? value)
    {
        var sorted = SortKeysRecursive(value);
        return JsonSerializer.Serialize(sorted);
    }

    private static object? SortKeysRecursive(object? value)
    {
        if (value is Dictionary<string, object?> dict)
        {
            var sorted = new SortedDictionary<string, object?>(StringComparer.Ordinal);
            foreach (var kvp in dict)
            {
                sorted[kvp.Key] = SortKeysRecursive(kvp.Value);
            }
            return sorted;
        }
        if (value is List<object?> list)
        {
            return list.Select(SortKeysRecursive).ToList();
        }
        return value;
    }

    private static bool EnumContains(List<object?> values, object? candidate)
    {
        foreach (var v in values)
        {
            if (v is null && candidate is null)
            {
                return true;
            }
            if (v is string vs && candidate is string cs && vs == cs)
            {
                return true;
            }
            if (v is bool vb && candidate is bool cb && vb == cb)
            {
                return true;
            }
            if (v is long || v is int || v is double || v is float)
            {
                if (candidate is long || candidate is int || candidate is double || candidate is float)
                {
                    var a = JsonUtil.AsDouble(v!);
                    var b = JsonUtil.AsDouble(candidate!);
                    if (Math.Abs(a - b) < 1e-12)
                    {
                        return true;
                    }
                }
            }
        }
        return false;
    }
}

public static class ValidatorUtil
{
    public static List<ValidationError> Validate(
        object? instance,
        TypeDef targetType,
        Schema schema,
        Registry registry,
        bool failFast = false)
    {
        return new Validator(registry, failFast).Validate(instance, targetType, schema);
    }
}

File: java/src/main/java/org/oojs/Validator.java
package org.oojs;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.oojs.model.*;

import java.util.*;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/** Stateful OOJS instance validator. */
public class Validator {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Registry registry;
    private final boolean failFast;
    private Map<String, Map<String, Object>> graphIndex = null;

    public Validator(Registry registry) {
        this(registry, false);
    }

    public Validator(Registry registry, boolean failFast) {
        this.registry = registry;
        this.failFast = failFast;
    }

    // ------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------

    public List<ValidationError> validate(Object instance, TypeDef targetType, Schema schema) {
        List<ValidationError> errors = new ArrayList<>();
        validateInstance(instance, targetType, schema, "/", errors);
        return errors;
    }

    public List<ValidationError> validateJson(String text, String targetTypeName, Schema schema) {
        Object instance;
        try {
            instance = MAPPER.readValue(text, Object.class);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid JSON: " + e.getMessage());
        }
        TypeDef typedef = schema.types.get(targetTypeName);
        if (typedef == null) {
            throw new IllegalArgumentException("Type '" + targetTypeName + "' not found in schema '" + schema.schemaId + "'");
        }
        return validate(instance, typedef, schema);
    }

    @SuppressWarnings("unchecked")
    public List<ValidationError> validateGraphDocument(Map<String, Object> graphDoc, Schema schema) {
        List<ValidationError> errors = new ArrayList<>();
        String discName = schema.discriminator;

        // Build $id -> raw-object index from roots and objects.
        Map<String, Map<String, Object>> index = new LinkedHashMap<>();
        Object rawRoots = graphDoc.get("roots");
        if (rawRoots instanceof List) {
            List<?> roots = (List<?>) rawRoots;
            for (Object rootObj : roots) {
                if (!(rootObj instanceof Map)) {
                    continue;
                }
                Map<String, Object> root = (Map<String, Object>) rootObj;
                Object idObj = root.get("$id");
                if (idObj instanceof String) {
                    index.put((String) idObj, root);
                }
            }
        }
        Object rawObjects = graphDoc.get("objects");
        if (rawObjects instanceof Map) {
            Map<?, ?> objects = (Map<?, ?>) rawObjects;
            for (Map.Entry<?, ?> entry : objects.entrySet()) {
                if (!(entry.getValue() instanceof Map)) {
                    continue;
                }
                index.put(String.valueOf(entry.getKey()), (Map<String, Object>) entry.getValue());
            }
        }
        graphIndex = index;

        Map<String, Map<String, Object>> positions = new LinkedHashMap<>();
        if (rawRoots instanceof List) {
            List<?> roots = (List<?>) rawRoots;
            for (int i = 0; i < roots.size(); i++) {
                if (roots.get(i) instanceof Map) {
                    positions.put("roots/" + i, (Map<String, Object>) roots.get(i));
                }
            }
        }
        if (rawObjects instanceof Map) {
            Map<?, ?> objects = (Map<?, ?>) rawObjects;
            for (Map.Entry<?, ?> entry : objects.entrySet()) {
                if (entry.getValue() instanceof Map) {
                    positions.put("objects/" + escape(String.valueOf(entry.getKey())),
                            (Map<String, Object>) entry.getValue());
                }
            }
        }

        for (Map.Entry<String, Map<String, Object>> pos : positions.entrySet()) {
            String path = pos.getKey();
            Map<String, Object> raw = pos.getValue();

            // Convert graph-object shape to regular instance shape.
            Map<String, Object> instance = new LinkedHashMap<>();
            for (Map.Entry<String, Object> entry : raw.entrySet()) {
                String key = entry.getKey();
                if ("$id".equals(key)) {
                    continue;
                }
                if ("$type".equals(key)) {
                    instance.put(discName, entry.getValue());
                    continue;
                }
                instance.put(key, entry.getValue());
            }

            Object discVal = instance.get(discName);
            if (!(discVal instanceof String)) {
                errors.add(new ValidationError(path, ErrorCode.MISSING_DISCRIMINATOR,
                        "graph object missing $type field"));
                continue;
            }
            TypeDef typedef = registry.lookupByDiscriminatorValue((String) discVal);
            if (typedef == null) {
                errors.add(new ValidationError(path + "/$type", ErrorCode.UNKNOWN_TYPE,
                        "unknown type '" + discVal + "'"));
                continue;
            }

            validateInstance(instance, typedef, schema, path, errors);
        }

        graphIndex = null;
        return errors;
    }

    // ------------------------------------------------------------------
    // Core validation phases (§9.2)
    // ------------------------------------------------------------------

    private boolean validateInstance(Object instance, TypeDef target, Schema schema,
            String path, List<ValidationError> errors) {
        if (!(instance instanceof Map)) {
            errors.add(new ValidationError(path, ErrorCode.TYPE_MISMATCH, "expected a JSON object"));
            return !failFast;
        }
        @SuppressWarnings("unchecked")
        Map<String, Object> obj = (Map<String, Object>) instance;

        // Phase 1: discriminator resolution
        String discName = schema.discriminator;
        if (!obj.containsKey(discName)) {
            errors.add(new ValidationError(path, ErrorCode.MISSING_DISCRIMINATOR,
                    "missing discriminator property '" + discName + "'"));
            return !failFast;
        }
        Object discVal = obj.get(discName);
        if (!(discVal instanceof String)) {
            errors.add(new ValidationError(path + "/" + discName, ErrorCode.INVALID_DISCRIMINATOR_TYPE,
                    "discriminator property '" + discName + "' must be a string"));
            return !failFast;
        }
        String discStr = (String) discVal;
        TypeDef concrete = registry.lookupByDiscriminatorValue(discStr);
        if (concrete == null) {
            errors.add(new ValidationError(path + "/" + discName, ErrorCode.UNKNOWN_TYPE,
                    "unknown type '" + discStr + "'"));
            return !failFast;
        }
        if (concrete.abstractType) {
            errors.add(new ValidationError(path + "/" + discName, ErrorCode.ABSTRACT_TYPE,
                    "type '" + discStr + "' is abstract and cannot be instantiated"));
            return !failFast;
        }
        if (!concrete.isSubtypeOf(target)) {
            errors.add(new ValidationError(path + "/" + discName, ErrorCode.TYPE_MISMATCH,
                    "type '" + discStr + "' is not a subtype of '" + target.name + "'"));
            return !failFast;
        }

        // Phase 2: required properties
        for (String reqName : concrete.effectiveRequired()) {
            if (!obj.containsKey(reqName)) {
                errors.add(new ValidationError(path, ErrorCode.MISSING_REQUIRED,
                        "missing required property '" + reqName + "'"));
                if (failFast) return false;
            }
        }

        // Phase 3: property presence + validation
        Map<String, PropertyDef> effectiveProps = concrete.effectiveProperties();
        for (Map.Entry<String, Object> entry : obj.entrySet()) {
            String key = entry.getKey();
            if (key.equals(discName)) continue;
            if (!effectiveProps.containsKey(key)) {
                if (schema.closedWorld) {
                    errors.add(new ValidationError(path + "/" + escape(key), ErrorCode.ADDITIONAL_PROPERTY,
                            "unexpected additional property '" + key + "'"));
                    if (failFast) return false;
                }
                continue;
            }
            boolean ok = validateProperty(entry.getValue(), effectiveProps.get(key), schema,
                    path + "/" + escape(key), errors);
            if (!ok && failFast) return false;
        }

        return true;
    }

    private boolean validateProperty(Object value, PropertyDef prop, Schema schema,
            String path, List<ValidationError> errors) {
        if (prop instanceof ArrayProperty) return validateArray(value, (ArrayProperty) prop, schema, path, errors);
        if (prop instanceof PrimitiveProperty) return validatePrimitive(value, (PrimitiveProperty) prop, path, errors);
        if (prop instanceof TypeRefProperty) return validateTypeRef(value, (TypeRefProperty) prop, schema, path, errors);
        return true;
    }

    // -- Array (§9.2 Phase 4, array branch) ------------------------------

    private boolean validateArray(Object value, ArrayProperty prop, Schema schema,
            String path, List<ValidationError> errors) {
        if (!(value instanceof List)) {
            errors.add(new ValidationError(path, ErrorCode.TYPE_MISMATCH, "expected an array"));
            return !failFast;
        }
        List<?> list = (List<?>) value;
        int count = list.size();

        if (count < prop.minItems) {
            errors.add(new ValidationError(path, ErrorCode.ARRAY_TOO_SHORT,
                    "array has " + count + " item(s), minimum is " + prop.minItems));
            if (failFast) return false;
        }
        if (prop.maxItems != null && count > prop.maxItems) {
            errors.add(new ValidationError(path, ErrorCode.ARRAY_TOO_LONG,
                    "array has " + count + " item(s), maximum is " + prop.maxItems));
            if (failFast) return false;
        }
        if (prop.uniqueItems) {
            Set<String> seen = new LinkedHashSet<>();
            for (Object item : list) {
                String key = serialiseForUniqueness(item);
                if (!seen.add(key)) {
                    errors.add(new ValidationError(path, ErrorCode.ARRAY_DUPLICATE_ITEMS, "array items must be unique"));
                    if (failFast) return false;
                    break;
                }
            }
        }
        for (int i = 0; i < list.size(); i++) {
            boolean ok = validateProperty(list.get(i), prop.items, schema, path + "/" + i, errors);
            if (!ok && failFast) return false;
        }
        return true;
    }

    // -- Primitive (§9.3) ------------------------------------------------

    private boolean validatePrimitive(Object value, PrimitiveProperty prop,
            String path, List<ValidationError> errors) {
        if (!jsonKindMatches(value, prop.kind)) {
            errors.add(new ValidationError(path, ErrorCode.TYPE_MISMATCH,
                    "expected " + prop.kind + ", got " + jsonTypeName(value)));
            return !failFast;
        }

        if ("integer".equals(prop.kind) && value instanceof Double) {
            double d = (Double) value;
            if (d != Math.floor(d) || Double.isInfinite(d)) {
                errors.add(new ValidationError(path, ErrorCode.NOT_INTEGER, "value must be an integer (no fractional part)"));
                if (failFast) return false;
            }
        }

        if ("string".equals(prop.kind)) {
            String s = (String) value;
            int length = s.codePointCount(0, s.length());
            if (prop.minLength != null && length < prop.minLength) {
                errors.add(new ValidationError(path, ErrorCode.STRING_TOO_SHORT,
                        "string length " + length + " < minLength " + prop.minLength));
                if (failFast) return false;
            }
            if (prop.maxLength != null && length > prop.maxLength) {
                errors.add(new ValidationError(path, ErrorCode.STRING_TOO_LONG,
                        "string length " + length + " > maxLength " + prop.maxLength));
                if (failFast) return false;
            }
            if (prop.pattern != null) {
                try {
                    if (!Pattern.compile(prop.pattern).matcher(s).find()) {
                        errors.add(new ValidationError(path, ErrorCode.PATTERN_MISMATCH,
                                "value does not match pattern '" + prop.pattern + "'"));
                        if (failFast) return false;
                    }
                } catch (PatternSyntaxException e) {
                    errors.add(new ValidationError(path, ErrorCode.PATTERN_MISMATCH,
                            "invalid pattern '" + prop.pattern + "': " + e.getMessage()));
                    if (failFast) return false;
                }
            }
            if (prop.enumValues != null && !prop.enumValues.contains(value)) {
                errors.add(new ValidationError(path, ErrorCode.ENUM_MISMATCH,
                        "value '" + value + "' not in enum " + prop.enumValues));
                if (failFast) return false;
            }
        } else if ("integer".equals(prop.kind) || "number".equals(prop.kind)) {
            double num = ((Number) value).doubleValue();
            if (prop.minimum != null && num < prop.minimum) {
                errors.add(new ValidationError(path, ErrorCode.BELOW_MINIMUM,
                        "value " + value + " < minimum " + prop.minimum));
                if (failFast) return false;
            }
            if (prop.maximum != null && num > prop.maximum) {
                errors.add(new ValidationError(path, ErrorCode.ABOVE_MAXIMUM,
                        "value " + value + " > maximum " + prop.maximum));
                if (failFast) return false;
            }
            if (prop.exclusiveMinimum != null && num <= prop.exclusiveMinimum) {
                errors.add(new ValidationError(path, ErrorCode.BELOW_EXCLUSIVE_MINIMUM,
                        "value " + value + " must be > " + prop.exclusiveMinimum));
                if (failFast) return false;
            }
            if (prop.exclusiveMaximum != null && num >= prop.exclusiveMaximum) {
                errors.add(new ValidationError(path, ErrorCode.ABOVE_EXCLUSIVE_MAXIMUM,
                        "value " + value + " must be < " + prop.exclusiveMaximum));
                if (failFast) return false;
            }
            if (prop.multipleOf != null) {
                double remainder = Math.abs(num) % prop.multipleOf;
                double rounded = Math.round(remainder * 1e10) / 1e10;
                if (rounded != 0.0 && rounded != prop.multipleOf) {
                    errors.add(new ValidationError(path, ErrorCode.NOT_MULTIPLE_OF,
                            "value " + value + " is not a multiple of " + prop.multipleOf));
                    if (failFast) return false;
                }
            }
            if (prop.enumValues != null) {
                boolean found = false;
                for (Object ev : prop.enumValues) {
                    if (ev instanceof Number && ((Number) ev).doubleValue() == num) { found = true; break; }
                }
                if (!found) {
                    errors.add(new ValidationError(path, ErrorCode.ENUM_MISMATCH,
                            "value " + value + " not in enum " + prop.enumValues));
                    if (failFast) return false;
                }
            }
        }
        return true;
    }

    // -- Type reference (§9.2 Phase 4, type-ref branch) ------------------

    private boolean validateTypeRef(Object value, TypeRefProperty prop, Schema schema,
            String path, List<ValidationError> errors) {
        // Graph-document mode: { "$ref-id": "..." } references another top-level
        // graph object. We type-check the referenced object's $type and avoid
        // recursive inline validation so cycles terminate naturally.
        if (graphIndex != null && value instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> maybeRef = (Map<String, Object>) value;
            if (maybeRef.size() == 1 && maybeRef.containsKey("$ref-id")) {
                Object refIdObj = maybeRef.get("$ref-id");
                if (!(refIdObj instanceof String)) {
                    errors.add(new ValidationError(path, ErrorCode.TYPE_MISMATCH, "$ref-id must be a string"));
                    return !failFast;
                }
                String refId = (String) refIdObj;
                Map<String, Object> target = graphIndex.get(refId);
                if (target == null) {
                    errors.add(new ValidationError(path, ErrorCode.UNRESOLVED_REFERENCE,
                            "unresolved $ref-id '" + refId + "'"));
                    return !failFast;
                }
                Object targetTypeNameObj = target.get("$type");
                if (!(targetTypeNameObj instanceof String)) {
                    errors.add(new ValidationError(path, ErrorCode.MISSING_DISCRIMINATOR,
                            "referenced object missing $type"));
                    return !failFast;
                }
                String targetTypeName = (String) targetTypeNameObj;
                TypeDef targetTypedef = registry.lookupByDiscriminatorValue(targetTypeName);
                TypeDef refTypedef = prop.resolvedType;
                if (targetTypedef == null || refTypedef == null) {
                    errors.add(new ValidationError(path, ErrorCode.UNKNOWN_TYPE,
                            "cannot resolve type '" + prop.typeName + "'"));
                    return !failFast;
                }
                if (!targetTypedef.isSubtypeOf(refTypedef)) {
                    errors.add(new ValidationError(path, ErrorCode.TYPE_MISMATCH,
                            "type '" + targetTypeName + "' is not a subtype of '" + prop.typeName + "'"));
                    return !failFast;
                }
                return true;
            }
        }

        if (!(value instanceof Map)) {
            errors.add(new ValidationError(path, ErrorCode.TYPE_MISMATCH, "expected a JSON object for type reference"));
            return !failFast;
        }

        // resolvedType is guaranteed non-null by eager registry resolution (§A.3).
        // A null here indicates a schema that bypassed the registry load path.
        TypeDef refTypedef = prop.resolvedType;
        if (refTypedef == null) {
            errors.add(new ValidationError(path, ErrorCode.UNKNOWN_TYPE,
                    "type reference '" + prop.typeName + "' was not resolved at load time"));
            return !failFast;
        }

        Schema refSchema = registry.getSchema(refTypedef.schemaId);
        if (refSchema == null) refSchema = schema;
        return validateInstance(value, refTypedef, refSchema, path, errors);
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private static boolean jsonKindMatches(Object value, String kind) {
        switch (kind) {
            case "null":    return value == null;
            case "boolean": return value instanceof Boolean;
            case "string":  return value instanceof String;
            case "integer": return value instanceof Integer || value instanceof Long || value instanceof Double;
            case "number":  return value instanceof Integer || value instanceof Long || value instanceof Double;
            default:        return false;
        }
    }

    private static String jsonTypeName(Object value) {
        if (value == null)             return "null";
        if (value instanceof Boolean)  return "boolean";
        if (value instanceof String)   return "string";
        if (value instanceof Integer || value instanceof Long) return "integer";
        if (value instanceof Double || value instanceof Float) return "number";
        if (value instanceof List)     return "array";
        if (value instanceof Map)      return "object";
        return value.getClass().getSimpleName();
    }

    private static String escape(String token) {
        return token.replace("~", "~0").replace("/", "~1");
    }

    private static String serialiseForUniqueness(Object value) {
        try {
            return MAPPER.writeValueAsString(sortKeysRecursive(value));
        } catch (JsonProcessingException e) {
            return String.valueOf(value);
        }
    }

    @SuppressWarnings("unchecked")
    private static Object sortKeysRecursive(Object value) {
        if (value instanceof Map) {
            Map<String, Object> sorted = new TreeMap<>();
            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
                sorted.put(String.valueOf(entry.getKey()), sortKeysRecursive(entry.getValue()));
            }
            return sorted;
        }
        if (value instanceof List) {
            List<Object> result = new ArrayList<>();
            for (Object item : (List<?>) value) result.add(sortKeysRecursive(item));
            return result;
        }
        return value;
    }
}
